From 52a26bb98cdc129e985294e984e362fe663f2637 Mon Sep 17 00:00:00 2001
From: Baokun Li <libaokun1@huawei.com>
Date: Wed, 5 Jul 2023 14:31:20 +0800
Subject: [PATCH 774/780] [Huawei] Revert "[Huawei] quota: fix race condition
 between dqput() and dquot_mark_dquot_dirty()"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188812

--------------------------------

This reverts commit 457bf4b45100aad605486141e257ead74e8570c8.
Revert self-research solution to synchronize the community has reviewed
solution.

Signed-off-by: Baokun Li <libaokun1@huawei.com>
---
 fs/quota/dquot.c | 23 ++++-------------------
 1 file changed, 4 insertions(+), 19 deletions(-)

diff --git a/fs/quota/dquot.c b/fs/quota/dquot.c
index cde0f4e3ae54..ad255f8ab5c5 100644
--- a/fs/quota/dquot.c
+++ b/fs/quota/dquot.c
@@ -364,26 +364,11 @@ int dquot_mark_dquot_dirty(struct dquot *dquot)
 		return 1;
 
 	spin_lock(&dq_list_lock);
-	ret = test_and_set_bit(DQ_MOD_B, &dquot->dq_flags);
-	if (ret)
-		goto out_lock;
-	spin_unlock(&dq_list_lock);
-
-	/*
-	 * Wait for dq_lock - after this we know that either dquot_release() is
-	 * already finished or it will be canceled due to DQ_MOD_B flag test.
-	 */
-	wait_on_dquot(dquot);
-	spin_lock(&dq_list_lock);
-	if (!test_bit(DQ_ACTIVE_B, &dquot->dq_flags)) {
-		clear_bit(DQ_MOD_B, &dquot->dq_flags);
-		goto out_lock;
-	}
-	/* DQ_MOD_B is cleared means that the dquot has been written back */
-	if (test_bit(DQ_MOD_B, &dquot->dq_flags))
+	if (!test_and_set_bit(DQ_MOD_B, &dquot->dq_flags)) {
 		list_add(&dquot->dq_dirty, &sb_dqopt(dquot->dq_sb)->
 				info[dquot->dq_id.type].dqi_dirty_list);
-out_lock:
+		ret = 0;
+	}
 	spin_unlock(&dq_list_lock);
 	return ret;
 }
@@ -808,7 +793,7 @@ void dqput(struct dquot *dquot)
 		return;
 	}
 	/* Need to release dquot? */
-	if (test_bit(DQ_ACTIVE_B, &dquot->dq_flags) && dquot_dirty(dquot)) {
+	if (dquot_dirty(dquot)) {
 		spin_unlock(&dq_list_lock);
 		/* Commit dquot before releasing */
 		ret = dquot->dq_sb->dq_op->write_dquot(dquot);
-- 
2.17.1

